kythe/typescript/testdata/class.ts

```diff
@@ -52,6 +67,10 @@ class Class implements IFace {
         this.member;
         //- @method ref Method
         this.method();
+        //- @mem ref GetMember
+        this.mem;
+        //- @mem ref SetMember
+        this.mem = 0;
```

Comment:
What happens in code like


```
class C {
  get x(): string { return 'a' }
}

interface I {
  x: string;
}

let o: I = new C();
o.x;
```

Where o.x refers to a getter but you can't "see" it from the usage site?
I think maybe o.x refers to I.x, and then we lose the connection between I.x and C.x#getter?

Reply:
Yep, `o.x` will refer to I.x. The same thing happens if `x` is a property or function in `class C`. It's a difficult to tell what the desired behavior here is since `o` has a type of I but a value of C. For me either resolution is acceptable, but I'll defer to your judgement.

Reply:
It's possible to have different representations of the same thing, but I recommend trying to avoid it. When some of the uses do xref with the accessors, and some of the uses do not, users will notice and will definitely complain.

There is no universal solution, but one option is indirection: For example, you could squint and pretend that the explicitly-defined accessors are overrides of implicitly-defined default ones. In that model, any assignment to a field is a ref/call to the implicit setter, and that will cluster with the explicit one via override resolution. (I'm not saying this is the _right_ solution, but I wanted to give an example of how to think about these issues).

Reply:
> @creachadair is this in regards to this interface/class example or the getter/setter idea as a whole?

It's in response to the example @evmar raised, that the indexer might not be able to tell, from the usage site `o.x` whether it dispatches to a getter (say) or is a direct access to the field.

The resulting cross-references can be confusing or even misleading if they're mixed: A developer may infer that the getter isn't being used and try to issue a fix, only to discover the bug wasn't in the code but in the index. With good tests no harm is done, but it's a waste of their time and effort. (This example may sound hypothetical, but it's a distillation of real bug reports we had on Grok over the years)

Conversely, if the indexer speculatively assumes there is always a getter, but there actually _is_ no getter, those cross-references won't link up to the field. 

So despite seeming fairly simple, this is actually an interesting design question for the indexer.

Reply:
> Conversely, if the indexer speculatively assumes there is always a getter, but there actually is no getter, those cross-references won't link up to the field.

Right, I see what you're saying. The indexer assumes nothing about where identifiers are defined; it defers entirely to the TypeScript compiler which is producing the behavior here (I wish I could provide a direct link, but you can plug this into https://ts-ast-viewer.com to see).

> the indexer might not be able to tell, from the usage site o.x whether it dispatches to a getter (say) or is a direct access to the field.

> you could squint and pretend that the explicitly-defined accessors are overrides of implicitly-defined default ones.

If the desire is to pick one behavior, I think this is the better one. We can find the most-overidden (is that the right term?) accessor, but not always the implicit one:

```typescript
class C {
  get x(): string { return 'a' }
}

interface I {
  x: string;
}

let o = new C();
o.x;
```

There is now no relationship between C and I, and there is no way of knowing the definition in `I` exists.

Reply:
I realized that in my original example, we already do the "wrong" thing, regardless of this change, because of exactly the example @ayazhafiz provided at the end here -- in a structural type system, you cannot ever "know" all the ways in which types interact.  So forgive my red herring there.

With some more thought, I think my objection here is that in code like
```
o.x = 3;
```
what this code is doing is "peeking in" to the definition of `o` to see whether it has a setter or not.

But per our separate compilation model, you only can peek in to files in the same compilation unit as the current file, and our compilation model is to have many small units.

So I think I've convinced myself that we should always emit the same thing on prop access, which means that for setters/getters we should just name them the same as the property.  Does that make sense?

Reply:
> So I think I've convinced myself that we should always emit the same thing on prop access, which means that for setters/getters we should just name them the same as the property. Does that make sense?

It does. There's a couple of caveats here - the signature of the getter and setter will have to be different (`foo` for the getter and `foo#setter` for the setter, because otherwise there would be two identical VNames; this still means that if there is only a setter its signature won't be the same as a property, but that is a concern only for the VName specification), and the reference emitted will be the first declaration discovered. We can ask that this points to the getter, but if only a setter is defined it will have to point to that. So in code like

```typescript
class A {
  get o() {...}
  set o(...) {...}
```

`A.o`  can always reference the getter, but in 

```typescript
class B {
  set o(...) {...}
```

`B.o` can of course only reference the setter.

Reply:
The play is now
1. `getter`s refer to themselves
2. `setter`s refer to themselves
3. references to a property that _has a getter_ (with or without a setter) refers to the getter
4. references to a property that _only has a setter_ refers to the setter